File: artifacts/api-server/src/routes/files.ts
import { Router, type IRouter } from "express";
import multer from "multer";
import { sql, eq, and, count } from "drizzle-orm";
import { db, documents, documentChunks, tenants } from "../lib/db";
import { requireJwt } from "../middlewares/auth";
import { extractText, chunkText } from "../lib/parser";
import { embedTexts, toVectorLiteral } from "../lib/embeddings";
import { fileOut } from "../lib/serialize";

const upload = multer({ storage: multer.memoryStorage(), limits: { fileSize: 20 * 1024 * 1024 } });
const router: IRouter = Router();

router.get("/files", requireJwt, async (req, res) => {
  const tenantId = req.auth!.tid;
  const tenantKey = req.auth!.tk;
  const docs = await db.select().from(documents).where(eq(documents.tenantId, tenantId));
  const counts = new Map<number, number>();
  if (docs.length > 0) {
    const rows = await db
      .select({ documentId: documentChunks.documentId, n: count() })
      .from(documentChunks)
      .where(eq(documentChunks.tenantId, tenantId))
      .groupBy(documentChunks.documentId);
    for (const r of rows) counts.set(r.documentId, Number(r.n));
  }
  return res.json(docs.map((d) => fileOut(d, tenantKey, counts.get(d.id) ?? 0)));
});

async function processAndStore(
  tenantId: number,
  filename: string,
  mimetype: string,
  buffer: Buffer,
): Promise<{ documentId: number; chunkCount: number; status: string; type: string }> {
  const parsed = await extractText(buffer, filename, mimetype);
  const chunks = chunkText(parsed.text, 800);
  const [doc] = await db
    .insert(documents)
    .values({
      tenantId,
      name: filename,
      type: parsed.type,
      sizeBytes: buffer.length,
      status: chunks.length > 0 ? "indexed" : "empty",
    })
    .returning();

  if (chunks.length === 0) {
    return { documentId: doc.id, chunkCount: 0, status: "empty", type: parsed.type };
  }

  const embeddings = await embedTexts(chunks);
  for (let i = 0; i < chunks.length; i++) {
    const emb = embeddings[i];
    if (emb) {
      await db.execute(sql`
        INSERT INTO document_chunks (tenant_id, document_id, chunk_index, content, embedding)
        VALUES (${tenantId}, ${doc.id}, ${i}, ${chunks[i]}, ${toVectorLiteral(emb)}::vector)
      `);
    } else {
      await db.insert(documentChunks).values({
        tenantId,
        documentId: doc.id,
        chunkIndex: i,
        content: chunks[i],
      });
    }
  }
  return { documentId: doc.id, chunkCount: chunks.length, status: "indexed", type: parsed.type };
}

// Canonical upload endpoint. Both web and mobile call this path through
// their shared api.ts helper.
router.post("/files/upload", requireJwt, upload.single("file"), async (req, res) => {
  const file = (req as unknown as { file?: Express.Multer.File }).file;
  if (!file) return res.status(400).json({ detail: "No file uploaded" });
  const tenantId = req.auth!.tid;
  try {
    const result = await processAndStore(tenantId, file.originalname, file.mimetype, file.buffer);
    return res.json({ file: file.originalname, chunks: result.chunkCount, status: result.status });
  } catch (err) {
    return res.status(500).json({ detail: (err as Error).message || "Upload failed" });
  }
});

router.delete("/files/:id", requireJwt, async (req, res) => {
  const tenantId = req.auth!.tid;
  const id = Number(req.params.id);
  if (Number.isNaN(id)) return res.status(400).json({ detail: "Bad id" });
  const [existing] = await db
    .select()
    .from(documents)
    .where(and(eq(documents.id, id), eq(documents.tenantId, tenantId)))
    .limit(1);
  if (!existing) return res.status(404).json({ detail: "Not found" });
  // Cascade fires on chunks (FK onDelete: cascade), no orphan vectors left.
  await db.delete(documents).where(eq(documents.id, id));
  return res.json({ ok: true });
});

export { processAndStore };
export default router;

// Silence unused-import warning when re-exported.
void tenants;

File: artifacts/api-server/app/routes/files.py
from typing import Annotated
from fastapi import APIRouter, Depends, HTTPException, UploadFile, File
from sqlalchemy import func, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.auth import get_current_user
from app.database import get_db
from app.models import Document, DocumentChunk
from app.schemas import FileOut
from app.services.embeddings import embed_texts
from app.services.parser import extract_text, chunk_text
from app.services.serialize import humanize_bytes, iso

router = APIRouter()


def _file_out(doc: Document, tenant_key: str, chunk_count: int) -> FileOut:
    return FileOut(
        id=doc.id,
        tenant_key=tenant_key,
        name=doc.name,
        size=humanize_bytes(doc.size_bytes),
        type=doc.type,
        chunks=chunk_count,
        date=iso(doc.created_at),
        status=doc.status,
    )


@router.get("/files", response_model=list[FileOut])
async def list_files(
    auth: Annotated[dict, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Document).where(Document.tenant_id == auth["tid"]).order_by(Document.created_at.desc())
    )
    docs = result.scalars().all()

    out = []
    for doc in docs:
        count_result = await db.execute(
            select(func.count(DocumentChunk.id)).where(DocumentChunk.document_id == doc.id)
        )
        chunk_count = count_result.scalar_one() or 0
        out.append(_file_out(doc, auth["tk"], chunk_count))
    return out


@router.post("/upload")
async def upload_file(
    auth: Annotated[dict, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db),
    file: UploadFile = File(...),
):
    content = await file.read()
    if len(content) > 20 * 1024 * 1024:
        raise HTTPException(status_code=400, detail="File too large (max 20MB)")

    text, file_type = await extract_text(content, file.filename or "upload", file.content_type or "")
    chunks = chunk_text(text)

    status = "indexed" if chunks else "empty"
    doc = Document(
        tenant_id=auth["tid"],
        name=file.filename or "upload",
        type=file_type,
        size_bytes=len(content),
        status=status,
    )
    db.add(doc)
    await db.flush()

    if chunks:
        embeddings = await embed_texts(chunks)
        for i, (chunk, emb) in enumerate(zip(chunks, embeddings)):
            db.add(DocumentChunk(
                tenant_id=auth["tid"],
                document_id=doc.id,
                chunk_index=i,
                content=chunk,
                embedding=emb,
            ))

    await db.commit()
    return {"file": file.filename, "chunks": len(chunks), "status": status}


@router.delete("/files/{file_id}")
async def delete_file(
    file_id: int,
    auth: Annotated[dict, Depends(get_current_user)],
    db: AsyncSession = Depends(get_db),
):
    result = await db.execute(
        select(Document).where(Document.id == file_id, Document.tenant_id == auth["tid"])
    )
    doc = result.scalar_one_or_none()
    if not doc:
        raise HTTPException(status_code=404, detail="File not found")
    await db.delete(doc)
    await db.commit()
    return {"ok": True}
